### PR TITLE
Feature/sg 636 pose estimation metrics

### PR DESCRIPTION
**Pull request by BloodAxe**

### CODE REVIEW

File: src/super_gradients/training/metrics/pose_estimation_metrics.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+                crowd_targets_areas=crowd_targets_areas,
+                crowd_targets_bboxes=crowd_targets_bboxes,
+                crowd_targets_ignored=crowd_targets_ignored,
+                #
+                iou_thresholds=self.iou_thresholds,
+                sigmas=self.oks_sigmas,
+                top_k=self.max_objects_per_image,
+            )
+        except IndexError as e:
+            raise e
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_metrics.py:218]: Is there an advantage of catching it if you raise it with no change right after?

File: src/super_gradients/training/metrics/pose_estimation_metrics.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        if self.rank is None:
+            self.rank = torch.distributed.get_rank() if self.is_distributed else -1
+
+        if self.is_distributed:
+            local_state_dict = self.predictions
+            gathered_state_dicts = [None] * self.world_size
+            torch.distributed.barrier()
+            torch.distributed.all_gather_object(gathered_state_dicts, local_state_dict)
+            self.predictions = list(itertools.chain(*gathered_state_dicts))
+            # matching_info = tensor_container_to_device(matching_info, device="cpu" if self.accumulate_on_cpu else self.device)
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_metrics.py:240]: This should be removed I guess

File: src/super_gradients/training/metrics/pose_estimation_utils.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    targets: torch.Tensor,
+    targets_visibilities: torch.Tensor,
+    targets_areas: Optional[torch.Tensor],
+    targets_bboxes: Optional[torch.Tensor],
+    targets_ignored: Optional[torch.Tensor],
+    crowd_targets: torch.Tensor,
+    crowd_visibilities: torch.Tensor,
+    crowd_targets_areas: Optional[torch.Tensor],
+    crowd_targets_bboxes: Optional[torch.Tensor],
+    crowd_targets_ignored: Optional[torch.Tensor],
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_utils.py:111]: I would add a default value of None for the Optional params

File: src/super_gradients/training/metrics/pose_estimation_utils.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    :param crowd_targets:   crowd targets for all images of shape (total_num_crowd_targets, 6)
+                            format:     (index, x, y, w, h, label) where x,y,w,h are in range [0,1]
+    :param iou_thresholds:  Threshold to compute the mAP
+    :param top_k:           Number of predictions to keep per class, ordered by confidence score
+
+    :return:
+        :preds_matched:     Tensor of shape (num_img_predictions, n_iou_thresholds)
+                                True when prediction (i) is matched with a target with respect to the (j)th IoU threshold
+        :preds_to_ignore:   Tensor of shape (num_img_predictions, n_iou_thresholds)
+                                True when prediction (i) is matched with a crowd target with respect to the (j)th IoU threshold
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_utils.py:130]: Only some parameters are described.

File: tests/unit_tests/pose_estimation_metrics_test.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                        "bbox": list([left_top[0], left_top[1], w, h]),
+                    }
+                )
+
+        return cat_results
+
+    def _coco_process_keypoints(self, keypoints):
+        tmp = keypoints.copy()
+        if keypoints[:, 2].max() > 0:
+            # p = keypoints[keypoints[:, 2] > 0][:, :2].mean(axis=0) ## TODO: It was unused variable
```
Comment (Louis-Dupont) [inline tests/unit_tests/pose_estimation_metrics_test.py:272]: Should this be removed?

File: src/super_gradients/training/metrics/pose_estimation_metrics.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        If you need to get additional metrics (AP at specific threshold), pass these thresholds via `iou_thresholds_to_report` argument.
+
+        :param post_prediction_callback:
+        :param num_joints:
+        :param oks_sigmas: OKS sigma factor for custom keypoint detection dataset.
+                           If None, then metric will use default OKS from COCO and expect num_joints to be equal 17
+        :param recall_thresholds: List of recall thresholds to compute AP.
+                                  If None, then will use default 101 recall thresholds from COCO in range [0..1]
+        :param iou_thresholds: List of IoU thresholds to use. If None, then COCO version of IoU will be used (0.5 ... 0.95)
+        :param: iou_thresholds_to_report: List of IoU thresholds to return in metric.
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_metrics.py:47]: `max_objects_per_image` is not described

File: src/super_gradients/training/metrics/pose_estimation_metrics.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/metrics/pose_estimation_metrics.py]: No, that's a leftover. Thanks for spotting this)

File: src/super_gradients/training/metrics/pose_estimation_metrics.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        num_joints: int,
+        max_objects_per_image: int = 20,
+        oks_sigmas: Optional[Iterable] = None,
+        iou_thresholds: Optional[Iterable] = None,
+        recall_thresholds: Optional[Iterable] = None,
+        iou_thresholds_to_report: Optional[Iterable] = None,
+    ):
+        """
+        Compute the AP & AR metrics for pose estimation. By default, this class returns only AP and AR values.
+        If you need to get additional metrics (AP at specific threshold), pass these thresholds via `iou_thresholds_to_report` argument.
```
Comment (Louis-Dupont) [inline src/super_gradients/training/metrics/pose_estimation_metrics.py:38]: The purpose is clear, but I think it would be easier to use if we also add a short example showing what the compute returns, especially when iou_thresholds_to_report is not None.

Something like:
```
>>> metric = PoseEstimationMetrics (iou_thresholds_to_report = [0.5, 0.9], ...)
>>> metric.update(...)
>>> metric.compute()
{"AP": ..., "AR":..., "AP_0.5": ..., "AR_0.5": ..., "AP_0.9": ..., "AR_0.9": ...}
```

File: src/super_gradients/training/metrics/pose_estimation_metrics.py (hunk 6 above)
Comment (BloodAxe) [reply src/super_gradients/training/metrics/pose_estimation_metrics.py]: Nice catch!

File: src/super_gradients/training/metrics/pose_estimation_metrics.py (hunk 7 above)
Comment (BloodAxe) [reply src/super_gradients/training/metrics/pose_estimation_metrics.py:53]: That is a great suggestion. Added

**Review by shaydeci — APPROVED** [review]
LGTM
